### PR TITLE
Paper: Developing an LSTM Pipeline for Accelerometer data

### PR DESCRIPTION
**Pull request by xtianmcd**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

### CODE REVIEW

**Comment by deniederhut** [comment]
Two thousand files seems a bit heavy for one paper. It looks like some of these are your tex source files. Could we ask you to remove them from your branch such that only the paper and associated data are left?

**Comment by xtianmcd** [comment]
My apologies. The problem should be fixed. Please let me know if there are still issues. I am having trouble getting the citations from my .bib file to be recognized in the built pdf on the build server… I have included the anchor in the .rst file as the example shows, and cited all citations with :cite:`citation_key`. Any advice?

**Comment by deniederhut** [comment]
@mpacer advice for failing bibfiles?

File: papers/christian_mcdaniel/citations.bib [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+@Inproceedings{WuAdu2017,
+  author = "Elias Wu and Paa Adu",
+  year = "2017",
+  title = "What Am I Doing? Robust Human Activity Detection with Smartphones Athletics & Sensing Devices",
+}
+
+@Article{Riveraetal2017,
+  author = "Patricio Rivera, et. al.",
+  year = "2017",
+  title "Recognition of Human Hand Activities Based on a Single Wrist IMU Using Recurrent Neural Networks",
```
Comment (mpacer) [inline papers/christian_mcdaniel/citations.bib]: needs an `=` sign to be valid.

File: papers/christian_mcdaniel/citations.bib [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+  title = "Multimodal Gesture Recognition Using 3-D Convolution and Convolutional LSTM",
+  publisher = "IEEE Access Volume: 5",
+  pages = "4517 - 4524"
+  doi = "10.1109/ACCESS.2017.2684186",
+}
+
+@Inproceedings{WuAdu2017,
+  author = "Elias Wu and Paa Adu",
+  year = "2017",
+  title = "What Am I Doing? Robust Human Activity Detection with Smartphones Athletics & Sensing Devices",
```
Comment (mpacer) [inline papers/christian_mcdaniel/citations.bib]: you likely need to escape `&`

**Comment by mpacer** [comment]
@xtianmcd Does the paper build locally? I ask in part because it doesn't build locally for me and I found some errors in your bibfile just glancing through it. (I made some comments above)

You might want to try running it through a bibfile validator like https://github.com/Pezmc/BibLatex-Check (though I don't know if that works only on BibLaTeX  files which are technically different…). But I would definitely try to find a bibfile validator of some kind to help you hunt down why your bibfile isn't working.

**Comment by xtianmcd** [comment]
Hi @mpacer I do get a local build (pdf output that matches the pdf in the build server), I also made corrections to my .bib file suggested by the BibLatex-Check and it returns no more errors. However, both my local build and the build on the server show [?] for my in-text citations and no References section.

**Comment by nicholasmalaya** [comment]
Hi @xtianmcd, I just pulled down a version of your paper to review from http://procbuild.scipy.org/# -- I'll try to get you comments ASAP, with a full review by June 21st. If major revisions occur it might make sense to note it here so I and the other reviewers take note.

**Comment by nicholasmalaya** [comment]
Hi @xtianmcd, 

I finally read this in detail. A few major and a few trivial comments:

Major: 
- Is the code publicly available? I don't see a public repo referenced on the paper, and I think this would really help the paper. You have gone through a great deal of effort in part because you claim other results in the field are not reproducible with respect to parameter selection and code traceability. I would really like to see this paper provide a high standard in this regard. Several ideas on this front:
   - Can you provide pointers to the code and data? Github or other public repos would be preferred. 
   - What about a unit/regression test that ensures the output is 'close' (by some metric) to the results in the paper?
   - I believe a table (or summary) should be provided that completely defines the network. I am not convinced that I could entirely reproduce your network architecture given this paper. Some of this information may be provided but is opaque (why units={{choise(numpy.arange(2,522,20))}}, instead of defining the number of layers and units in a table or text, for example).
   - One could provide a docker image or container that very simply provides a means to grab this result 
- I'm concerned this paper assumes too much familiarity with LSTMs (or RNNs in general). As this is a survey paper intended to in part bridge the reproducability gap, I think adding a short section at the beginning defining precisely what you are using (e.g., the logic gates, activations, etc.) in a mathematical sense should be useful. This would likely serve as a good LSTM intro to SciPy attendees. 

Trivialities: 
- It occurs to me that I could more easily provide these in-line, and will add them to the text.

**Comment by xtianmcd** [comment]
@nicholasmalaya thanks so much for your review. I will definitely post my code to GitHub; I will construct a table for the network; and I will include a more detailed background regarding RNN and LSTM. Should have all this done by some time next week; I'll update this thread when done. 

Side note: I am still working to get the build process to incorporate my .bib file... I have used BibLatex-Check referenced by @mpacer above and no issues are detected in my . bib file, but it will not build the references section or insert the in-text citations as far as I can tell. Any ideas? Worst case I can just add the citations in directly to .rst file

**Comment by nicholasmalaya** [comment]
> Any ideas? Worst case I can just add the citations in directly to .rst file

That is a good contingency plan, but let me take a look at building this tomorrow and see if I can detect anything. .bib debugging can be quite subtle, but I picked up a few tricks when I wrote my thesis :P 

I'll update if I find anything.

**Comment by nicholasmalaya** [comment]
Just an FYI: I've now captured all my comments on the paper at this time.

**Comment by rsivalingam** [comment]
Hi @xtianmcd 
The paper is well written and comprehensive. Please find my comments below.
- IN the code snippet on Page 4, column 2, when optimizer is passed in as "adam", "rmsprop", and "sgd" as inside quotes, they would end up using the default versions of those optimizers rather than those initialized in the previous lines. Could you please confirm and update the lines if the pre-initialized optimizers were indeed used?

- A table compiling all the results and comparisons would give a concise representation and easy for the reader to quickly take in.

- Will the code for undoing the pre-processing of the UCI HAR dataset also be released, along with the LSTM network training and testing code?

Thank you
Ravi

**Comment by xtianmcd** [comment]
Thank you for your comments @rsivalingam. 

I had left out a few lines (that allow to me to tune the learning rate) from the code snippet for readability; however, I am working to add in more code to the paper and will include this portion with a short explanation as well. 

I am also creating a table for the results/comparisons and the architecture/pipeline. Should have these done shortly after I finish the Background section with some math/theory of LSTMs. 

Yes, I will be releasing all code shortly including all pre-processing/undoing of pre-processing on the data. 

Thanks again, I will notify here when updates are submitted.

**Comment by deniederhut** [comment]
Hi @xtianmcd ! How are those updates going?

**Comment by deniederhut** [comment]
On a related note, it looks like the server is having trouble compiling your paper. Can you confirm that it builds locally?

**Comment by xtianmcd** [comment]
Hi @deniederhut the updates are coming along well; I am almost done with them. I am working on the paper currently and some of the changes I just made are causing build issues so I'm troubleshooting through them. I should have it resolved shortly though and will make a PR with all my changes so far once its ability to build is restored.

**Comment by xtianmcd** [comment]
I have made the PR, so you all should be able to view my updates so far @deniederhut @rsivalingam @nicholasmalaya. 

Further work:
1) I have one or two more tables to add and 2) it needs to be shortened further to get it under 9pgs once 3) the references are added (which I still cannot get to compile from .bib file; may just add in manually). 

Please let me know of any further revisions, suggestions or feedback!

**Comment by deniederhut** [comment]
It's finding the bibfile okay, but there are typos in it so it's not compiling. I might go through and look for missing commas, unmatched quotations marks, and unescaped ampersands. Judging by what I have locally, there are also some citations keys in your paper that don't match the ones in the bibfile.

**Comment by deniederhut** [comment]
@nicholasmalaya @rsivalingam do you feel this paper is ready to be included in the proceedings?

**Comment by rsivalingam** [comment]
Just downloaded the latest version - will have my comments by EOD Monday

**Comment by nicholasmalaya** [comment]
@xtianmcd this is clearly much improved and addresses several of the comments from previously. Thank you for your efforts! 

One minor comment: you mentioned that you need this down to 8 pages, but I don't believe that includes references. I'm only seeing 8 pages of writing, so you might be OK on that front.

Additional comments:
* TABLE 1appears to be bleeding into the other column
* Can you make the font in figures 1,3,4 larger? It is hard to read
* You have a typo, 'Baye’s Theorem'-->"Bayes Theorem"

@deniederhut once the comments above and the bibliograph are fixed I would recommend this paper for publication.

**Comment by xtianmcd** [comment]
@nicholasmalaya thanks for your feedback. I'll get working on these comments now. That's good to hear re: the length. 

When I view the PDF from the build server, it looks like the References section is now building, but in-text citations are all still "[?]". I'm hoping they will be recognized if I change all citation keys to the auto-generated ones next to each citation. Will report back today.

**Comment by deniederhut** [comment]
Anything with a .bib file needs to be compiled twice. So if you run `make papers` locally two times in a row, you'll see the correct references in your local copy.

Sorry that it doesn't show up on procbuild.

**Comment by xtianmcd** [comment]
@deniederhut Hmm, when I build locally I actually do not see the references section still. 

Also, there is a discrepancy with the table formats. When I build locally, the tables look fine now, but when I view on procbuild the tables are not formatted correctly anymore. This also causes the lengths of the paper to be different - should I trust the local build? 

Lastly, I just updated the title of the paper; I want to make sure it is ok to do so at this point.

**Comment by deniederhut** [comment]
Okay, so I haven't fetched your paper since 90c2337ed86b194e679e83d0d7a975d885f98761, but I can get a local build like this:

screenshot 2018-07-03 18 05 14
...
screenshot 2018-07-03 18 05 27

**Comment by deniederhut** [comment]
First guess is that your build is being polluted by an old aux file. Try:

```sh
make clean
make papers
make papers
```

**Comment by xtianmcd** [comment]
Hi @deniederhut @nicholasmalaya @rsivalingam I have just submitted my final PR. However, I still cannot view the in-text citations from my local build for some reason. Would one of you mind building the paper locally and confirming that the in-text citations are working properly. With that confirmed, my paper will be ready for submission!

**Comment by deniederhut** [comment]
> I still cannot view the in-text citations from my local build for some reason

I would look at the contents of `output/christian_mcdaniel/paper.log`. If rebuilding the aux file didn't solve the issue, my next guess is that there is a typo in one of your entries that is causing bibtex to crash.

One more thing -- papers should not modify anything outside of their own directories. You have two `.DS_Store` files that you've committed to other parts of the repo. Please remove these.

**Comment by xtianmcd** [comment]
I can now see the in-text citations perfectly in the `procbuild` build server! 

Once the remaining changes are Verified, the paper is ready to go. All in-text citations will work and the length should be within the limit!

**Comment by deniederhut** [comment]
Hooray! The server doesn't pull down new revisions immediately - I would
give it a half hour and check again.

**Comment by xtianmcd** [comment]
The changes haven't been Verified yet but I'm hoping that since my paper is ready and I've received a recommendation for approval that my paper will be approved for inclusion. Let me know if there's anything I can do to ensure this happens. Thanks!

**Comment by deniederhut** [comment]
Hey @rsivalingam ! I think we're still waiting on your approval for this paper.

**Comment by rsivalingam** [comment]
@xtianmcd 
The paper looks greatly improved - nice work! Citations work for me now, and the new figures/tables are very helpful. 
I approve this paper, pending a few very minor things to fix below:

- Fix typo in Fig. 1 - "n: (n)um Attibutes = 3"
- In the equation in the background section: 
--Is the p in yp supposed to mean "predicted"? Please specify that. Or I recommend using y with a tilde on top, as is common in ML literature.
--If you are doing transpose, you don't need to use the dot product - it is redundant - you can simply say w . x, that alleviates the question of whether w is a row or column vector
- At muliple places, please correct "backpropogation" to "backpropagation"
- Fig 4 D - 659200 or 65900? It is different in different places
- Pg 6, last line before Conclusion/Future Work - OVO SVM - can you specify that OVO is one-vs-one, and maybe add a reference? This may not be obvious to all readers.

**Comment by xtianmcd** [comment]
@rsivalingam thank you very much for your revisions! I have made the changes and submitted the PR. Once the build server incorporates the changes, the paper will be ready to go :)

**Comment by xtianmcd** [comment]
Thank you so much @deniederhut @rsivalingam @nicholasmalaya for your help and patience getting this paper finished! Hope I get to meet all of you in the coming days :)

**Comment by xtianmcd** [comment]
hi @deniederhut I don't see the paper on the proceedings site http://conference.scipy.org/proceedings/scipy2018/. Was it able to be included?

**Comment by deniederhut** [comment]
Yup! Just not right this minute. The papers you see posted are the papers that were ready by the deadline.

**Comment by xtianmcd** [comment]
Wonderful, thanks :)
